magda-io/magda
# Issue/2778 Make Magda helm charts v3 charts & migration docs

### What this PR does

Fixes #2778 

> Merged documentation PR: https://github.com/magda-io/magda/pull/2801 Please merge it first

Summary:
- Make Magda Helm chart requires Helm v3 & Kube API >= 1.14
- Fixed helm repo index generation using Helm 3 binary
- Added [v0.0.56-RC6-to-v0.0.57-0.md](https://github.com/magda-io/magda/blob/8979f9e2322c2d94b0c0dcc893f190680667fa35/docs/docs/migration/v0.0.56-RC6-to-v0.0.57-0.md) doc
- Also, make sure `magda-config` users can deploy with terraform  and create PR here:
https://github.com/magda-io/magda-config/pull/10

### Test Repo index generation script

Before starting the test, install helm 2 (v2.16.3) using home brew and make the binary is accessible as `helm2` (helm v3 is accessible through helm)

Test Repo index generation script with local repo:
- Check out this PR branch `issue/2778`
- Delete the following directories files (if exists):
   - `deploy/helm/magda/charts`
   - `deploy/helm/magda/requirements.lock`
   - `deploy/helm/magda/requirements.yaml`
   - `deploy/helm/magda-core/charts`
   - `deploy/helm/magda-core/requirements.lock`
   - `deploy/helm/magda-core/requirements.yaml`
   - `deploy/helm/internal-charts/storage-api/charts`
   - `deploy/helm/internal-charts/storage-api/requirements.lock`
   - `deploy/helm/internal-charts/storage-api/requirements.yaml`
- Set AWS s3 access key & secret in env
- Run the followings:
```bash
cd deploy/helm
# we are going to serve local repo from this dir
mkdir chart-repo 
# download magda-charts bucket content
aws s3 sync s3://magda-charts/ chart-repo
mkdir -p sync_dir
mkdir -p index_dir
# download existing index file for index merging
if ! aws s3 cp s3://magda-charts/index.yaml index_dir/index.yaml; then echo "failed to copy current repo index" && exit 1; fi
helm package -d sync_dir openfaas
helm package -d sync_dir magda-core
helm package -d sync_dir magda
# generate index with merging
helm repo index --merge "index_dir/index.yaml" sync_dir
cp -f sync_dir/* chart-repo
```

Start local chart repo:
```bash
yarn global add local-web-server
cd chart-repo
ws
```

The local chart repo will be serving at http://127.0.0.1:8000

You can deploy Magda v0.0.56-RC6 with [v0.0.56-RC6 version `minikube-dev.yml`](https://github.com/magda-io/magda/blob/v0.0.56-RC6/deploy/helm/minikube-dev.yml) (with docker image change to `docker hub`)

```bash
helm2 --namespace magda upgrade --install --timeout 9999 --wait -f minikube-dev.yml magda magda --repo http://127.0.0.1:8000 --version 0.0.56-RC6
```

###  Test Migration Locally with Minikube

Download `helm_deployment_migration.sh` script from: https://gist.github.com/t83714/9317f8933223d3c4bd626128ee306208

Changes line 193 
from:
`helm get manifest "${helm_deployment_name}"`
to:
`helm2 get manifest "${helm_deployment_name}"`
as helm2 should be accessible via `helm2` for our test.

Run the following to build & push 0.0.57-0 docker images to local minikube (if you haven't done so):
```
yarn lerna run build --stream --concurrency=1 --include-dependencies
eval $(minikube docker)
yarn lerna run docker-build-local --stream --concurrency=1 --include-filtered-dependencies
```

Follow the instruction document [v0.0.56-RC6-to-v0.0.57-0.md](https://github.com/magda-io/magda/blob/768718d3a87bb33f9d0a4dc24fc08b51e3a65d79/docs/docs/migration/v0.0.56-RC6-to-v0.0.57-0.md) for migration except `Step 6`.

You can use the following to install the chart from your local repo instead:
```bash
helm --namespace magda upgrade --install --timeout 9999s --wait -f deploy/helm/minikube-dev.yml magda magda --repo http://127.0.0.1:8000 --version 0.0.57-0 --set global.connectors.includeInitialJobs=false
```

### Checklist

-   [x] unit tests aren't applicable
